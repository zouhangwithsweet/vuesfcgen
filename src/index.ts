import { baseParse, TemplateChildNode } from '@vue/compiler-core'
import { transform } from '@babel/core'

const NORMAL_ATTR = 6

// babel transform script code
function transformScript(ast: TemplateChildNode, options?: any) {
  if ('tag' in ast) {
    if (ast.tag === 'script') {
      const sourceCode = ast.loc.source
      let code = ``
      code += transform(sourceCode, options)
      return code
    }
  }
  return ``
}

// todo transform style ast
function transformStyle(ast: TemplateChildNode, options?: any) {
  if ('tag' in ast) {
    if (ast.tag === 'style') {
      return ast.loc.source
    }
  }
  return ``
}

// gen template
function createCodegenContext(ast: TemplateChildNode) {
  const context = {
    code: ``,
    push(code: string) {
      context.code += code
    },
    source: ast.loc.source,
    newline(n = 2) {
      context.push('\n' + ` `.repeat(n))
    }
  }
  return context
}

function genTemplate(ast: TemplateChildNode) {
  const context = createCodegenContext(ast)
  const { push, newline } = context

  push('<template>')
  newline()

  genHTML(ast, context)

  push('</template>')
  newline()
  return context
}

function genHTML(ast: TemplateChildNode, context) {
  const { push, newline } = context
  if ('children' in ast) {
    ast.children.forEach(child => {
      const { tag, isSelfClosing, props } = child
      if (tag) {
        if (isSelfClosing) {
          push(`<${tag}`)
          injectProps(props, context)
          push(`/>`)
          newline()
        } else {
          push(`<${tag}`)
          injectProps(props, context)
          push(`>`)
          if (child.children && child.children.length) {
            genHTML(child, context)
          }
          push(`</${tag}>`)
          newline()
        }
      } else {
        push(child.loc.source)
      }
    })
  }
}

function injectProps(props, context) {
  const { push } = context
  props.forEach(prop => {
    push(` `)
    const { type, name, value, loc } = prop
    if (type === NORMAL_ATTR) {
      value
        ? push(`${name}=${value.loc.source}`)
        : push(`${name}`)
    } else {
      push(loc.source)
    }
  })
}
