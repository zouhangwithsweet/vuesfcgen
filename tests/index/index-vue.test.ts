import fs from 'fs'
import path from 'path'
import { mount } from '@vue/test-utils'
import Index from './Index.vue'
import { baseParse } from '@vue/compiler-core'
import { transformScript, transformStyle, genTemplate, createCodegenContext, formatCode } from '../../src/index'

test('default', () => {
  const wrapper = mount(Index)
  expect(wrapper.html()).toMatchSnapshot()
})

test('generator', async () => {
  const filePath = path.resolve(__dirname, './Index.vue')
  const content = fs.readFileSync(filePath).toString()
  const rootNode = baseParse(content)

  const scriptNode = rootNode.children.find(
    (_) => 'tag' in _ && _.tag === 'script',
  )
  const styleNode = rootNode.children.find(
    (_) => 'tag' in _ && _.tag === 'style',
  )
  const templateNode = rootNode.children.find(
    (_) => 'tag' in _ && _.tag === 'template',
  )

  if (!scriptNode || !styleNode || !templateNode) {
    throw new Error('Invalid Nodes')
  }

  const scriptResult = await transformScript(scriptNode, {
    filename: 'file.ts',
  })
  const styleResult = await transformStyle(styleNode)
  const templateResult = await genTemplate(templateNode)

  const context = createCodegenContext()
  context.push(templateResult)
  context.push('<script>')
  context.newline()
  context.push(scriptResult)
  context.newline()
  context.push('</script>')
  context.newline()
  context.push('<style>')
  context.newline()
  context.push(styleResult)
  context.newline()
  context.push('</style>')
  context.code = formatCode(context.code)

  const genFilePath = path.resolve(__dirname, './Index-gen.vue')
  const genComponent = require(genFilePath).default;
  const genWrapper = mount(genComponent)
  expect(genWrapper.html()).toMatchSnapshot()
})
