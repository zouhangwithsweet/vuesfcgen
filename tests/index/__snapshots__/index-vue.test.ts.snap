// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`default 1`] = `
"<div>
  <p aria-disabled=\\"\\" title=\\"Hello World!\\">Hello World!</p> <br>
  <h1></h1>
</div>"
`;

exports[`generator 1`] = `
"<div>
  <p aria-disabled=\\"\\" title=\\"Hello World!\\">
    Hello World!
  </p> <br>
  <h1></h1>
</div>"
`;
